chain/chain/src/chain.rs

```diff
@@ -1175,29 +1192,41 @@ impl Chain {
         Ok(!self.store.get_blocks_to_catchup(prev_prev_hash)?.contains(&prev_hash))
     }
 
+    /// Return all shards that whose states need to be caught up
+    /// That has two cases:
+    /// 1) Shard layout will change in the next epoch. In this case, the method returns all shards
+    ///    in the current epoch that will be split into a future shard that `me` will track.
+    /// 2) Shard layout will be the same. In this case, the method returns all shards that `me` will
+    ///    track in the next epoch but not this epoch
     fn get_shards_to_dl_state(
         &self,
         me: &Option<AccountId>,
         parent_hash: &CryptoHash,
     ) -> Vec<ShardId> {
-        let epoch_id = self.runtime_adapter().get_epoch_id_from_prev_block(parent_hash).unwrap();
+        let epoch_id = self.runtime_adapter.get_epoch_id_from_prev_block(parent_hash).unwrap();
         (0..self.runtime_adapter.num_shards(&epoch_id).unwrap())
             .filter(|shard_id| {
-                self.runtime_adapter.will_care_about_shard(
-                    me.as_ref(),
-                    parent_hash,
-                    *shard_id,
-                    true,
-                ) && !self.runtime_adapter.cares_about_shard(
-                    me.as_ref(),
-                    parent_hash,
-                    *shard_id,
-                    true,
-                )
+                Self::should_catch_up_shard(self.runtime_adapter(), me, parent_hash, *shard_id)
             })
             .collect()
     }
 
+    fn should_catch_up_shard(
+        runtime_adapter: Arc<dyn RuntimeAdapter>,
+        me: &Option<AccountId>,
+        parent_hash: &CryptoHash,
+        shard_id: ShardId,
+    ) -> bool {
+        let will_shard_layout_change =
+            runtime_adapter.will_shard_layout_change(parent_hash).unwrap();
```

Comment:
Same here

Reply:
Technically no because of the same reason I mentioned above, but I changed the code so that it doesn't just unwrap